updated Docker image conda environment.yml for osmnx=1.3.*

carlhiggs opened the pull request:
Updated Docker image conda environment.yml for osmnx=1.3.* (released Jan 2023) and sqlalchemy>=1.4,<2.0, with related code changes (mostly to address/avoid sqlalchemy deprecation warnings).   

As per #179 (which this pull request addresses), I wasn't able to successfully migrate all sqlalchemy-related code for the newer version 1.4,  which appeared with the updated dependencies arising from updated OSMnx.  This version prints to console verbose warnings in colour about impending changes in 2.0.  The difficulties in migration were partly because pandas and geopandas don't yet support sqlalchemy 2.0.   As such, I ultimately set the environment variable suggested by sqlalchemy for suppressing the not-yet-relevent deprecation warnings, as I couldn't address all of them.

The pull request also deals with the [dependabot alert for Pillow](https://github.com/global-healthy-liveable-cities/global-indicators/security/dependabot/12) >=9.2.0 ,<9.3.0  (updates to Pillow 9.4.0).

I haven't pushed the updated Docker image to DockerHub yet; I will do that after code review.

Incidentally I updated the template to include the new ghent_v3 entry, which displays some more complicated/complete study region area data sourcing than we've used elsewhere.  It queries a layer in geopackage by attributes to identify Ghent's municipal boundary.  This is just done via config file (passing ogr2ogr arguments for "-where"), not via code.  I thought it useful to retain as a usage example.

carlhiggs committed:
updated Docker image conda environment.yml for osmnx=1.3.* and sqlalchemy>=1.4,<2.0 as per #179, with related code changes (mostly to address/avoid sqlalchemy deprecation warnings)

carlhiggs commented:
I've done over-zealous tagging of reviewers, to compensate for not doing this on previous commits (sorry) --- trying to change ways!  Let's discuss protocol/preferred way of doing this in our next meeting perhaps :)

gboeing commented on docker/environment.yml:
```diff
@@ -3,8 +3,9 @@ channels:
   - conda-forge
 dependencies:
   - python=3.9.*
```
If we're upgrading the image anyway, is it worth moving to a more recent version of python as well? Most of the python scientific ecosystem uses the policy of supporting only the 3 most recent python versions. 3.9 is currently the third most recent version. Just a thought.

carlhiggs replied:
Good point Geoff, thanks - worth seeing if we can upgrade; I can look into this

gboeing commented on docker/requirements.txt:
```diff
@@ -1,6 +1,6 @@
 affine==2.3.1
-alabaster==0.7.12
-attrs==22.1.0
+appdirs==1.4.4
```
Out of curiosity, why pin these reqs all the way down to the patch version? Why not just major.minor?

carlhiggs replied:
Actually, there's been a change in the recent enhancement branch merge that simplified the Docker image - requirements.txt nows just a pip freeze of what is installed (maybe misnomer) for dependabot purposes.  The install was actually based on the much more sparse environment.yml.

Keen to here your thoughts on this approach, but there is less pinning in that file.

gboeing commented on process/pre_process/_project_setup.py:
```diff
@@ -205,6 +205,12 @@ def load_yaml(
+# and migration of syntax is not straightforward
+# (warnings are obscure); hence ignoring warnings
+# and pinning sqlalchemy < 2.0
+os.environ["SQLALCHEMY_SILENCE_UBER_WARNING"] = "1"
```
Seems perfectly reasonable for now.

carlhiggs committed:
Merge remote-tracking branch 'origin' into enhancements

carlhiggs committed:
updated environment.yml to require python=3.10.* (tried with 3.11.* unsuccessfully, dependencies not yet ready for newer version); also amended the study region creation wrapper script to not advise users to create validation report, as we are yet to update the template for this.  Thirdly, rearranged the tests in tests.py ahead of adding more tests

carlhiggs committed:
added in additional pre-commit hooks, and corrected if elif ... else logic in _project_setup.py for running the generate reports script

carlhiggs committed:
added in local running of unittest in pre-commit hooks (noting that as currently configured, this is not within Docker; may have to change implementation later if we require this)

carlhiggs committed:
added in local running of unittest in pre-commit hooks (noting that as currently configured, this is not within Docker; may have to change implementation later if we require this)

carlhiggs committed:
added in local running of unittest in pre-commit hooks (noting that as currently configured, this is not within Docker; may have to change implementation later if we require this)

carlhiggs committed:
added in local running of unittest in pre-commit hooks (noting that as currently configured, this is not within Docker; may have to change implementation later if we require this)

carlhiggs committed:
added in local running of unittest in pre-commit hooks (noting that as currently configured, this is not within Docker; may have to change implementation later if we require this)

carlhiggs committed:
added in local running of unittest in pre-commit hooks (noting that as currently configured, this is not within Docker; may have to change implementation later if we require this)

carlhiggs committed:
testing a change that should cause a unittest failure in the precommit hook to confirm that it does

carlhiggs committed:
testing a change that should cause a unittest failure in the precommit hook to confirm that it does

carlhiggs committed:
testing a change that should cause a unittest failure in the precommit hook to confirm that it does

carlhiggs committed:
renamed tests.py to move towards passing name-tests-test test

carlhiggs commented:
In the above I updated environment.yml to use Python 3.10.*  which I verified as working.  First I tried Python 3.11.* but this wouldn't resolve.

I did a whole load of commits testing out getting unittest to work as a local pre-commit; I couldn't really, so I removed that for now.  I can figure that out later (probably better to launch these within Docker anyway...).

@global-healthy-liveable-cities/software-spatial  I think this is ready to review now.   It incorporates the bug-fixes that were addressed in the main branch earlier today (including adding a test file), but also upgrades OSMnx to latest version, addresses a Pillow-related depend-a-bot alert, and updates Python.   Also added some more pre-commit hooks.

carlhiggs committed:
added in the updated pip freeze of requirements.txt (this is what happened to be installed; note that the docker image is built from environment.yml)

carlhiggs committed:
fixed up project setup check for test script (it was dependent on specific name for test script (tests.py); i made it work for any test script that starts with 'test_'

carlhiggs committed:
added back in a blank line inadvertently removed
